### ISSUE TITLE
__eq__ defined in baseclass inherited by frozen dataclass is not picked up and thus pyright says the return value is `false`

### ISSUE THREAD

**Issue by btruhand** [microsoft/pyright#9535]
**Describe the bug**
I have a similar code to the one below:

```python
from attr import dataclass


class Base:
    def __eq__(self, other):
        return Test(2)

@dataclass(frozen=True)
class Test(Base):
    a: int

def bar(x: Test):
    pass

b = Test(1) == Test(1)
print(bar(b))

```

The type of `b` will is considered to be `bool` in this case instead of `Test`

If the code is modified to any of the following:

1. Make `Test` not a dataclass or at least not a frozen dataclass
2. Define `__eq__` on `Test` rather than `Base`

then the type of `b` will correctly be inferred as `Test`

**Code or Screenshots**

Failing typecheck:
![Image](https://github.com/user-attachments/assets/20e2eafe-df90-495e-b117-3e0e23079554)

Passing typecheck:
![Image](https://github.com/user-attachments/assets/3bc83b7f-7fff-49fc-95b8-7d073214d764)

**VS Code extension or command-line**

VSCode version: 1.93.1
Pylance extension version: v2024.8.1
Python extension version: v2024.12.3

**Comment by erictraut** [comment]
Pyright is working correctly here.

If you insert a `print(type(b))` statement at the end of your sample and run the code, you'll see that the type of `b` is `bool`. That's because the `dataclass` decorator in `attr` works like the stdlib dataclass and synthesizes an `__eq__` method. If you don't want it to do this, you need to specify `eq=False` in the decorator.

```python
@dataclass(frozen=True, eq=False)
class Test(Base):
    a: int
```
